### DISCUSSION TITLE
"watch" mode

### DISCUSSION THREAD

**Discussion by benjie** [graphile/crystal-pre-merge#14] [Ideas]
V5 is split into these phases, effectively:

1. read/resolve the config
2. build the inflectors
3. perform the data gathering
4. generate the schema

Step 3 pulls data typically from external sources (e.g. a PostgreSQL database), and changes to that should trigger step 3 and 4 to run again. This is what V4 did and it's what V5 should do.

However, it'd be really nice if watch mode also applied to configuration/plugins/etc. To achieve that, however, is incredibly complex - we need to negotiate things like `require.cache` and asynchronous `import` without caches - and we need to do so without causing memory leaks.

We have similar concerns when doing watch mode in `graphile-worker`, and there's known bugs in the implementation we have for that.

For now, we're going to have watch mode just watch the PostgreSQL database I think - any changes to code (configuration, plugins, etc) should be handled by restarting the Node process via `nodemon` or similar.

If you have ideas on how to improve this situation, please detail them below!

**Comment by benjie** [comment]
Initial watch mode in #15

**Comment by ab-pm** [comment]
Hot module reloading doesn't appear to be very popular yet in node.js. There are a few projects though:

* https://github.com/fgnass/node-dev
* https://github.com/palmerhq/backpack
* https://github.com/jaredpalmer/razzle
* https://github.com/jlongster/backend-with-webpack
* https://hackernoon.com/hot-reload-all-the-things-ec0fed8ab0
